File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/truevfs/kernel/impl/NeedsSyncException.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.truevfs.kernel.impl

import de.schlichtherle.truecommons.shed._
import de.schlichtherle.truecommons.shed._
import javax.annotation.concurrent._

/** Indicates that a file system controller needs to get `sync`ed before the
  * operation can get retried.
  *
  * @see    FsSyncController
  * @author Christian Schlichtherle
  */
@Immutable
private final class NeedsSyncException private ()
extends ControlFlowException

private object NeedsSyncException {
  import ControlFlowException._

  private[this] val instance = new NeedsSyncException

  def apply() = if (isTraceable) new NeedsSyncException else instance
}

File: src/main/java/net/java/truevfs/kernel/impl/NeedsSyncException.java
/*
 * Copyright © 2005 - 2021 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.shed.BitField;
import net.java.truecommons.shed.ControlFlowException;

/**
 * Indicates that a file system controller needs to get
 * {@linkplain net.java.truevfs.kernel.spec.FsController#sync(BitField) synced} before the operation can get retried.
 *
 * @author Christian Schlichtherle
 * @see SyncController
 */
final class NeedsSyncException extends ControlFlowException {

    private static final long serialVersionUID = 0;
    private static final NeedsSyncException instance = new NeedsSyncException();

    private NeedsSyncException() {
        super(false);
    }

    static NeedsSyncException apply() {
        return isTraceable() ? new NeedsSyncException() : instance;
    }
}
